### PR TITLE
Less allocations for node processing

### PR DESCRIPTION
**Pull request by benaadams**
## Changes

- Introduce struct enumerators rather than using Linq

Addresses the following allocations
image

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Network.Discovery/RoutingTable/NodeTable.cs [hunk 1]
```diff
@@ -73,42 +76,140 @@ public void RefreshNode(Node node)
...
+
+    public ClosestNodesFromNodeEnumerator GetClosestNodes(byte[] nodeId, int bucketSize)
     {
         CheckInitialization();
+        return new ClosestNodesFromNodeEnumerator(Buckets, nodeId, _nodeDistanceCalculator, Math.Min(bucketSize, _discoveryConfig.BucketSize));
+    }
+
+    public struct ClosestNodesFromNodeEnumerator : IEnumerator<Node>, IEnumerable<Node>
+    {
+        private readonly List<Node> _sortedNodes;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network.Discovery/RoutingTable/NodeTable.cs:163]: ArrayPoolList?

File: src/Nethermind/Nethermind.Network.Discovery/RoutingTable/NodeBucket.cs [hunk 2]
```diff
@@ -28,26 +29,71 @@ public NodeBucket(int distance, int bucketSize, float dropFullBucketProbability
...
-                    node = node.Previous;
+                    return true;
                 }
             }
+
+            Current = null!;
+            return false;
+        }
+
+        public void Reset() => throw new NotSupportedException();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network.Discovery/RoutingTable/NodeBucket.cs:80]: Should we have Reset's throw not supported exceptions? Shouldn't be hard to reset all of them?

Comment (benaadams) [reply src/Nethermind/Nethermind.Network.Discovery/RoutingTable/NodeBucket.cs]: Not if never use it, docs say don't bother https://learn.microsoft.com/en-us/dotnet/api/system.collections.ienumerator.reset?view=net-8.0#remarks

image


Hid behind explicit implementation https://github.com/NethermindEth/nethermind/pull/6967/commits/e2586909ec940bb9271620645abedd6ddfe058a9

File: src/Nethermind/Nethermind.Network.Discovery/RoutingTable/NodeTable.cs (hunk 1 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Network.Discovery/RoutingTable/NodeTable.cs]: Changed to use ArrayPoolList https://github.com/NethermindEth/nethermind/pull/6967/commits/a144110a4b712760e32dabac8f9c60e2a758b9d2

File: src/Nethermind/Nethermind.Core/Collections/ArrayPoolList.cs [hunk 3]
```diff
@@ -121,6 +121,49 @@ void ICollection.CopyTo(Array array, int index)
...
+
+        Count = count;
+        if (count < _capacity / 2)
+        {
+            // Reduced to less than half of the capacity, resize the array.
+            T[] newArray = _arrayPool.Rent(count);
+            _array.AsSpan(0, count).CopyTo(newArray);
+            T[] oldArray = Interlocked.Exchange(ref _array, newArray);
+            _capacity = newArray.Length;
+            _arrayPool.Return(oldArray);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Collections/ArrayPoolList.cs:143]: Is it worth it, to potentially copy over this data to the new buffer? 
I expect this data to be consumed fairly quickly and the whole buffer would return.

IMO probably not worth it at

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Core/Collections/ArrayPoolList.cs (hunk 3 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Collections/ArrayPoolList.cs:143]: The consumers only want 12 or 16 elements by default; so copy isn't that big, but can be hundreds or thousands of nodes

**Pull request merged**